File: mobile/NoiseMonitor/app/src/main/java/com/example/noisemonitor/ui/HistoryFragment.java
package com.example.noisemonitor.ui;

import android.app.DatePickerDialog;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.noisemonitor.R;
import com.example.noisemonitor.api.ApiService;
import com.example.noisemonitor.api.HistoryEvent;
import com.google.android.material.switchmaterial.SwitchMaterial;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.List;
import java.util.Locale;

public class HistoryFragment extends Fragment {

    private RecyclerView recyclerView;
    private HistoryAdapter adapter;
    private final List<HistoryEvent> eventList = new ArrayList<>();
    private ApiService apiService;
    private ProgressBar progressBar;
    private SwitchMaterial criticalOnlySwitch;
    private ImageButton refreshButton;
    private Button todayButton, yesterdayButton, selectDateButton;

    private Calendar selectedDate = Calendar.getInstance();
    private final SimpleDateFormat apiDateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.getDefault());
    private boolean isInitialLoad = true;

    public HistoryFragment() {
        super(R.layout.fragment_history);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        apiService = ApiService.getInstance(requireContext());
        bindViews(view);
        setupRecyclerView();
        setupButtons();
    }

    @Override
    public void onResume() {
        super.onResume();
        isInitialLoad = true; // Reset on resume
        fetchHistoryData();
    }

    private void bindViews(@NonNull View view) {
        recyclerView = view.findViewById(R.id.recycler_view_history);
        progressBar = view.findViewById(R.id.progress_bar_history);
        criticalOnlySwitch = view.findViewById(R.id.switch_critical_only);
        refreshButton = view.findViewById(R.id.button_refresh);
        todayButton = view.findViewById(R.id.button_today);
        yesterdayButton = view.findViewById(R.id.button_yesterday);
        selectDateButton = view.findViewById(R.id.button_select_date);
    }

    private void setupRecyclerView() {
        LinearLayoutManager layoutManager = new LinearLayoutManager(requireContext());
        recyclerView.setLayoutManager(layoutManager);
        adapter = new HistoryAdapter(requireContext(), eventList);
        recyclerView.setAdapter(adapter);
        recyclerView.addItemDecoration(new DividerItemDecoration(requireContext(), layoutManager.getOrientation()));
    }

    private void setupButtons() {
        refreshButton.setOnClickListener(v -> fetchHistoryData());
        criticalOnlySwitch.setOnCheckedChangeListener((buttonView, isChecked) -> fetchHistoryData());

        todayButton.setOnClickListener(v -> {
            selectedDate = Calendar.getInstance();
            fetchHistoryData();
        });

        yesterdayButton.setOnClickListener(v -> {
            selectedDate = Calendar.getInstance();
            selectedDate.add(Calendar.DAY_OF_YEAR, -1);
            fetchHistoryData();
        });

        selectDateButton.setOnClickListener(v -> showDatePicker());
    }

    private void showDatePicker() {
        DatePickerDialog datePickerDialog = new DatePickerDialog(
                requireContext(),
                (datePicker, year, month, day) -> {
                    selectedDate.set(year, month, day);
                    fetchHistoryData();
                },
                selectedDate.get(Calendar.YEAR),
                selectedDate.get(Calendar.MONTH),
                selectedDate.get(Calendar.DAY_OF_MONTH)
        );
        datePickerDialog.show();
    }

    private void fetchHistoryData() {
        setLoading(true);
        String dateString = apiDateFormat.format(selectedDate.getTime());
        boolean criticalOnly = criticalOnlySwitch.isChecked();

        apiService.getHistoryEvents(dateString, criticalOnly, new ApiService.ApiCallback<List<HistoryEvent>>() {
            @Override
            public void onSuccess(List<HistoryEvent> result) {
                if (!isAdded()) return; 
                requireActivity().runOnUiThread(() -> {
                    setLoading(false);
                    eventList.clear();
                    eventList.addAll(result);
                    adapter.notifyDataSetChanged();
                    isInitialLoad = false; // Mark initial load as done
                    if (result.isEmpty()) {
                        Toast.makeText(getContext(), "No events found for this day.", Toast.LENGTH_SHORT).show();
                    }
                });
            }

            @Override
            public void onError(Exception e) {
                if (!isAdded()) return; 
                requireActivity().runOnUiThread(() -> {
                    setLoading(false);
                    isInitialLoad = false;
                    Toast.makeText(getContext(), "API Error: " + e.getMessage(), Toast.LENGTH_LONG).show();
                });
            }
        });
    }

    private void setLoading(boolean isLoading) {
        if (progressBar != null && recyclerView != null) {
            if (isLoading) {
                progressBar.setVisibility(View.VISIBLE);
                if (isInitialLoad) {
                    // On first load, hide the content to avoid showing old data
                    recyclerView.setVisibility(View.INVISIBLE);
                }
            } else {
                progressBar.setVisibility(View.GONE);
                // Always make sure content is visible when not loading
                recyclerView.setVisibility(View.VISIBLE);
            }
        }
    }
}

File: mobile/NoiseMonitor/app/src/main/java/com/example/noisemonitor/HistoryFragment.kt
package com.example.noisemonitor

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.launch
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.noisemonitor.databinding.FragmentHistoryBinding
import com.example.noisemonitor.network.RetrofitClient
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.core.content.ContextCompat
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

class HistoryFragment : Fragment() {

    private var _binding: FragmentHistoryBinding? = null
    private val binding get() = _binding!!

    private var selectedDate: String = getTodayDate()
    private var onlyCritical: Int = 0

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentHistoryBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        // 1️⃣ Настраиваем RecyclerView
        val layoutManager = LinearLayoutManager(requireContext())
        binding.measurementsRecycler.layoutManager = layoutManager
        
        // Добавляем разделитель
        val dividerItemDecoration = DividerItemDecoration(
            binding.measurementsRecycler.context,
            layoutManager.orientation
        )
        // Устанавливаем кастомный разделитель (drawable), который мы создали
        ContextCompat.getDrawable(requireContext(), R.drawable.divider_drawable)?.let {
            dividerItemDecoration.setDrawable(it)
        }
        
        binding.measurementsRecycler.addItemDecoration(dividerItemDecoration)

        binding.filterSwitch.setOnCheckedChangeListener { _, isChecked ->
            onlyCritical = if (isChecked) 1 else 0
            loadEvents()
        }

        binding.buttonToday.setOnClickListener {
            selectedDate = getTodayDate()
            loadEvents()
        }

        binding.buttonYesterday.setOnClickListener {
            selectedDate = getYesterdayDate()
            loadEvents()
        }

        binding.buttonPickDate.setOnClickListener {
            loadEvents()
        }

        binding.refreshButton.setOnClickListener {
            loadEvents()
        }

        loadEvents()
    }

    private fun loadEvents() {
        viewLifecycleOwner.lifecycleScope.launch {
            try {
                val events = RetrofitClient.api.getEvents(
                    date = selectedDate,
                    onlyCritical = onlyCritical
                )
                binding.measurementsRecycler.adapter = HistoryAdapter(events)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    private fun getTodayDate(): String {
        val sdf = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
        return sdf.format(Date())
    }

    private fun getYesterdayDate(): String {
        val sdf = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())
        val cal = Calendar.getInstance()
        cal.add(Calendar.DAY_OF_YEAR, -1)
        return sdf.format(cal.time)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}
